## Import Packages

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## 4.Using the pandas read_csv() method, read the GDP dataset into your notebook as a DataFrame called gdp_df.

In [4]:
pd.read_csv("../data/gdp_percapita.csv.csv")

Country or Area  Year        Value  Value Footnotes
0        Afghanistan  2021  1517.016266              NaN
1        Afghanistan  2020  1968.341002              NaN
2        Afghanistan  2019  2079.921861              NaN
3        Afghanistan  2018  2060.698973              NaN
4        Afghanistan  2017  2096.093111              NaN
...              ...   ...          ...              ...
7657        Zimbabwe  1994  2670.106615              NaN
7658        Zimbabwe  1993  2458.783255              NaN
7659        Zimbabwe  1992  2468.278257              NaN
7660        Zimbabwe  1991  2781.787843              NaN
7661        Zimbabwe  1990  2704.757299              NaN

[7662 rows x 4 columns]

In [5]:
gdp_df=pd.read_csv("../data/gdp_percapita.csv.csv")

## 5. How many rows and columns are in gdp_df? What are the data types of each column? 7662 rows x 4 columns

In [7]:
### gdp_df.shape

In [8]:
pd.read_csv("../data/gdp_percapita.csv.csv")
row_count = gdp_df.shape[0]
column_count = gdp_df.shape[1]
print(f'The DataFrame has {row_count} rows.')
print(f'The DataFrame as {column_count} rows.')

The DataFrame has 7662 rows.
The DataFrame as 4 rows.


## 6. Drop the Value Footnotes column and rename the remaining three to 'Country', 'Year', and 'GDP_Per_Capita'.


#### (A) Drop the Value Footnotes column:
##### df = df.drop('column_name', axis=1) 
###### -where 1 is the axis number (0 for rows and 1 for columns.)
##### **Dibran's alternative: gdp_df.drop(columns - ['Value Footnotes'])
##### **Mike's alternative: gdp_df.drop('column_name', axis='columns') 
##### To save to dataframe, gdp_df = 

In [11]:
gdp_df.drop('Value Footnotes', axis=1)

Country or Area  Year        Value
0        Afghanistan  2021  1517.016266
1        Afghanistan  2020  1968.341002
2        Afghanistan  2019  2079.921861
3        Afghanistan  2018  2060.698973
4        Afghanistan  2017  2096.093111
...              ...   ...          ...
7657        Zimbabwe  1994  2670.106615
7658        Zimbabwe  1993  2458.783255
7659        Zimbabwe  1992  2468.278257
7660        Zimbabwe  1991  2781.787843
7661        Zimbabwe  1990  2704.757299

[7662 rows x 3 columns]

#### (B)Rename the remaining three to 'Country', 'Year', and 'GDP_Per_Capita'.
##### gdp_df.rename(columns = {'x':'New name' 'y', 'New Name': 'z', 'New Name'})
##### **Dibran's alternative: df.columns = ['Country', 'Year', and 'GDP_Per_Capita']
###### Dibran's alternative requires renaming all columns (it cannot be used if not addressing every single column)

In [13]:
gdp_df.rename(columns = {'Country or Area': 'Country', 'Year': 'Year', 'Value': 'GDP_Per_Capita'})

Country  Year  GDP_Per_Capita  Value Footnotes
0     Afghanistan  2021     1517.016266              NaN
1     Afghanistan  2020     1968.341002              NaN
2     Afghanistan  2019     2079.921861              NaN
3     Afghanistan  2018     2060.698973              NaN
4     Afghanistan  2017     2096.093111              NaN
...           ...   ...             ...              ...
7657     Zimbabwe  1994     2670.106615              NaN
7658     Zimbabwe  1993     2458.783255              NaN
7659     Zimbabwe  1992     2468.278257              NaN
7660     Zimbabwe  1991     2781.787843              NaN
7661     Zimbabwe  1990     2704.757299              NaN

[7662 rows x 4 columns]

#### (C) Combine drop and rename functions:
##### df.rename(columns = {'x':'new_name', 'y':'new_name'}, inplace = False).drop('z',axis=1,inplace=False)
##### print(df)
##### ***Note: I was having trouble because I was using inplace = True, because that causes the first change to be saved. So repeating the function to make the change again go into error since it was already done

In [15]:
gdp_df.rename(columns = {'Country or Area': 'Country', 'Year': 'Year', 'Value': 'GDP_Per_Capita'}, inplace = False).drop('Value Footnotes', axis=1,inplace=False)

Country  Year  GDP_Per_Capita
0     Afghanistan  2021     1517.016266
1     Afghanistan  2020     1968.341002
2     Afghanistan  2019     2079.921861
3     Afghanistan  2018     2060.698973
4     Afghanistan  2017     2096.093111
...           ...   ...             ...
7657     Zimbabwe  1994     2670.106615
7658     Zimbabwe  1993     2458.783255
7659     Zimbabwe  1992     2468.278257
7660     Zimbabwe  1991     2781.787843
7661     Zimbabwe  1990     2704.757299

[7662 rows x 3 columns]

## 7. How many countries have data for all years? Which countries are missing many years of data? Look at the number of observations per year. What do you notice? 


#### (A) Find max and min years
#### gdp_df[''].max()
#### min_value = df['Year'].min()

In [18]:
max_value = gdp_df['Year'].max()
min_value = gdp_df['Year'].min()
print(f'The max year is {max_value}.')
print(f'The min year is {min_value}.')
difference = 2022 - 1990
print(difference) 

The max year is 2022.
The min year is 1990.
32


#### Count missing values in entire df
##### df.isna().sum().sum()
##### print(df)
#### Count missing values in each column
##### df.isna().sum()
###### Why is the above option not working? 

#### Dibran's answer:
#### ?(1) gdp_df['Country'].value_counts()
##### to check answer, gdp_df[gdp_df['Country']== 'Country name']
#### (2) Save to count_df

#### Countries that have data for all years
##### (3) count_df['count'] == 33
##### (4) count_df-count_df['count'] == 33].count()

#### Counties that have missing years:
##### (5) count_df-count_df['count'] < 33].count()

#### Observations
##### There will be a couple of countries that have more than 33 years

## 8.In this question, you're going to create some plots to show the distribution of GDP per capita for the year 2020. Go to the Python Graph Gallery (https://www.python-graph-gallery.com/) and look at the different types of plots under the Distribution section. Create a histogram, a density plot, a boxplot, and a violin plot. What do you notice when you look at these plots? How do the plots compare and what information can you get out of one type that you can't necessarily get out of the others?

In [22]:
### Narrow data to GDP per capita for 2020 and save as a new dataset

In [23]:
df_2020 = gdp_df[gdp_df['Year'] == 2020]

### Histogram

In [25]:
#### libraries & dataset
##### import seaborn as sns
##### import matplotlib.pyplot as plt
#### set a grey background (use sns.set_theme() if seaborn version 0.11.0 or above, set otherwise) 
##### sns.set_theme(style="darkgrid")
##### df = sns.load_dataset("iris")

##### sns.histplot(data=df, x="sepal_length", kde=True)
##### plt.show()

In [26]:
import seaborn as sns
import matplotlib.pyplot as plt
df_2020.histplot(data=df_2020, x="GDP_per_capita")
##### plt.show()
df_2020[].plot(kind='hist', edgecolor ='', bins =)

SyntaxError: invalid syntax (1740225032.py, line 5)

In [ ]:
#### Density plot

In [ ]:
sns.kdeplot

In [ ]:
#### Boxplot
sns.boxplot

In [ ]:
#### Violin plot

In [ ]:
sns.violinplot

## 9.What was the median GDP per capita value in 2020?

### Median function is median()
#### df_2020[GDP_Per_Capita'].median()

In [ ]:
df_2020['GDP_Per_Capita'].median()

## 10.Create some visualizations to compare GDP per capita values for the years 1990, 2000, 2010, and 2020. Start by subsetting your data to just these 4 years into a new DataFrame named gdp_decades. Comment on what you observe has happened to GDP values over time and the relative strengths and weaknesses of each type of plot. 
### A boxplot
### A barplot 
### A scatterplot
### A scatterplot with a trend line overlaid 

### Create subset for years 1990, 2000, 2010, and 2020
#### gdp_df[gdp_df['Year'].isin([1990,2000,2010,2020])]
#### Save above subset as a name (like gdp_decades)
##### Alternative: 
###### my_years = [1990,2000,2010,2020])
###### gdp_df[gdp_df['Year'].isin(my_years]


### Barplot 

### Boxplot

### Scatterplot

### Scatterplot with a trend line overlaid 

## 11. Which country was the first to have a GDP per capita greater than $100,000?

In [ ]:
gdp_df = gdp_df.rename(columns = {'Country or Area': 'Country', 'Year': 'Year', 'Value': 'GDP_Per_Capita'}, inplace = False).drop('Value Footnotes', axis=1,inplace=False)

In [ ]:
###  (A) Find countries with GDP per capita greater than $100,000.
#### gdp_df[gdp_df['GDP_Per_Capita']>100000]
### (B) Sort by year to find country with earliest year
#### sort_values(by='Year') 

In [ ]:
gdp_df[gdp_df['GDP_Per_Capita']>100000].sort_values(by='Year').head(1) 

## 12. Which country had the highest GDP per capita in 2020? Create a plot showing how this country's GDP per capita has changed over the timespan of the dataset.

### (A) Narrow to data for 2020: 
#### gdp_df[gdp_df['Year']==2020]

In [ ]:
gdp_df[gdp_df['Year']==2020]

### (B) Sort by GDP (descending):
#### sort_values('GDP_Per_Capita',ascending=True)
### (C) Narrow to top 1 using head(1):
#### sort_values('GDP_Per_Capita',ascending=True)
#### **Dibran's alternative: 
##### nlargest(1, 'GDP_Per_Capita')
#### ***Elizabeth's question: Can't max function be used? Yes, but would need to use & to do it in steps, because max will only return the largest GDP but without the country name.

In [ ]:
gdp_df[gdp_df['Year']==2020].sort_values(by='GDP_Per_Capita', ascending=False).head(1)

#### (D) Create dataset for Luxembourg and name it

In [ ]:
luxembourg = gdp_df[gdp_df['Country']=='Luxembourg']

### (E) Plot showing how this country's GDP per capita has changed over the timespan of the dataset.
#### .plot(x='',y='',kind='line')
##### .plot(x='Year',y='GDP_Per_Capita',kind='line')

In [ ]:
luxembourg.plot(title='Luxembourg', x='Year',y='GDP_Per_Capita',kind='line')

## 13. Which country had the lowest GDP per capita in 2020? Create a plot showing how this country's GDP per capita has changed over the timespan of the dataset. ** Bonus question: Is it true in general that countries had a higher GDP per capita in 2020 than in 1990? Which countries had lower GDP per capita in 2020 than in 1990?

### (A) GDP data in 2020: gdp_df[gdp_df['Year']==2020]
### (B) Lowest GDP per capita: sort_values(by='GDP_Per_Capita', ascending=True)
### (C) Obtain top 5 results using head()

In [ ]:
gdp_df[gdp_df['Year']==2020].sort_values(by='GDP_Per_Capita', ascending=True).head()

### (C) Create subset for Burundi
### (D) Create a plot showing how this country's GDP per capita has changed over the timespan.

In [ ]:
burundi = gdp_df[gdp_df['Country']=='Burundi']
print(burundi)

In [ ]:
burundi.plot(title='Burundi',x='Year',y='GDP_Per_Capita',kind='line')

## 14. Read in the internet use dataset into a DataFrame named internet_df. You will likely get errors when doing this. Check the arguments for the read_csv function to find ones that can help correct the errors (https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) Once you are able to read it in, take per a look at the top and bottom few rows to make sure that it has been read in correctly. Also, check the datatypes of the columns.

#### (A) Initial attempt: dropped columns: 
##### internet_df = pd.read_csv("../data/internet_use.csv.csv", usecols=[0,1,2,3])
###### --But footers actually need to be dropped (which will drop the rows with excess columns anyways)
#### (B) Final attempt: drop columns below 6084.
#####    1. Calculate number of columns: 6383
#####    2. Calculate number of Columns to drop: 6381-6082=299
#####    3.pd.read_csv("../data/internet_use.csv.csv", skipfooter=299) --> does not work unless in python engine
#### (C) Third attempt: Use first attempt to drop columns and then df =df[x;y] to include only the needed rows
#####    ---internetdropcol_df = pd.read_csv("../data/internet_use.csv.csv", usecols=[0,1,2,3])
#####    ---internet_df = pd.read_csv("../data/internet_use.csv.csv", [0:6083])
#####      - Error: SyntaxError: invalid syntax
#### (D Final attempt (which works!):  pd.read_csv("../data/internet_use.csv.csv",skiprows=list(range(6085,6385)))
#####---(Took some time to figure out parameters of rows to skip)

In [254]:
internet_df = pd.read_csv("../data/internet_use.csv.csv",skiprows=list(range(6085,6385)))

#### (1) Check data type: df.dtypes
##### Result: 4 columns, all with object data type

In [127]:
internet_df.dtypes

Country or Area     object
Year                 int64
Value              float64
Value Footnotes     object
dtype: object

#### (2) Check top and bottom rows: head() and tail ()
##### Top rows look fine and show reasonable data for Country or Area, Year, Value, and Value Footnotes
##### Bottom rows do not look the same-Country or Area is numbers, years appear to be notes, Value and Value Footnotes are NaN


In [130]:
internet_df.head()

Country or Area  Year  Value Value Footnotes
0     Afghanistan  2019  17.60               1
1     Afghanistan  2018  16.80               2
2     Afghanistan  2017  13.50               3
3     Afghanistan  2016  11.00               4
4     Afghanistan  2015   8.26               4

In [ ]:
## (not used, just an option df=pd.read_excel('Data_2.xlsx'))

In [132]:
internet_df.tail()

Country or Area  Year     Value Value Footnotes
6078        Zimbabwe  1997  0.033080             NaN
6079        Zimbabwe  1996  0.016790             NaN
6080        Zimbabwe  1995  0.007684             NaN
6081        Zimbabwe  1994  0.001739             NaN
6082        Zimbabwe  1990  0.000000             NaN

## 15. Drop the Value Footnotes column and rename the remaining three to 'Country', 'Year', and 'Internet_Users_Pct'.

#### (A) Drop Value Footnotes column: drop('z',axis=1,inplace=False)
#### (B) Rename the remaining three to 'Country', 'Year', and 'Internet_Users_Pct': df.rename(columns = {'x':'new_name', 'y':'new_name'}, inplace = False).
#### (C) Save as new subset 


In [258]:
internet_df = internet_df.drop('Value Footnotes', axis=1, inplace=False).rename(columns = {'Country or Area':'Country', 'Year':'Year', 'Value':'Internet_Users_Pct'}, inplace = False)

In [ ]:
## 16. Look at the number of observations in this dataset per year. What do you notice?

In [ ]:
#### (A) Group by year
#### (B) Count Internet_Users_Pct per year
#### (C) Count Countries per year group

## 17. What is the first year to have a non-zero internet users percentage value?

###  (A) Find Internet_Users_Pct that are greater than 0
#### internet_df[internet_df['Internet_Users_Pct']>0]
### (B) Sort by year to find country with earliest year
#### sort_values(by='Year') 

In [ ]:
#### Answer: Denmark, 1990, percentage = .097277

In [143]:
internet_df[internet_df['Internet_Users_Pct']>0].sort_values(by='Year')

Country  Year  Internet_Users_Pct
1570                 Denmark  1990            0.097277
2892                   Japan  1990            0.020294
306                Australia  1990            0.585095
2034                  France  1990            0.052778
339                  Austria  1990            0.130369
...                      ...   ...                 ...
427               Bangladesh  2023           44.502700
677   Bosnia and Herzegovina  2023           83.377200
1656                 Ecuador  2023           72.694300
1537                 Denmark  2023           98.775600
3057                  Kuwait  2023           99.747300

[5803 rows x 3 columns]

In [ ]:
## 18. How does the distribution of internet users percent differ for 2000 and 2014?

In [ ]:
#### (A) 2000_internet_df = internet_df[internet_df['Year']==2000]
#### (B) sns.distplot

In [288]:
internet_df['Year'] = pd.to_datetime(internet_df['Year'], format='%Y')

In [296]:
internet0014_df = pd.DataFrame({'Year':[2000,2014]})
print(internet0014_df)

   Year
0  2000
1  2014


In [222]:
internet2000_df = internet_df[internet_df['Year']==2000]
internet2014_df = internet_df[internet_df['Year']==2014]
sns.stripplot(data=internet0014merge_df, x="Year", y="Internet_Users_Pct")



                 Country  Year_x  Internet_Users_Pct_x  Year_y  \
0                Albania    2000              0.114097    2014   
1                Algeria    2000              0.491706    2014   
2                Andorra    2000             10.538800    2014   
3                 Angola    2000              0.105046    2014   
4               Anguilla    2000             22.405400    2014   
..                   ...     ...                   ...     ...   
190             Viet Nam    2000              0.254248    2014   
191  Virgin Islands (US)    2000             13.815100    2014   
192                Yemen    2000              0.082500    2014   
193               Zambia    2000              0.191072    2014   
194             Zimbabwe    2000              0.401434    2014   

     Internet_Users_Pct_y  
0                 54.3000  
1                 29.5000  
2                 86.1000  
3                 21.3623  
4                 70.3900  
..                    ...  
190        

In [ ]:
sns.displot(~ x,data=internet2000_df,y="Internet_Users_Pct", groups="Year")
sns.pairplot(data=internet2000_df,hue = 'Internet_Users_Pct',diag_kind = "kde",kind = "scatter",palette = "husl")
plt.show
sns.stripplot(data=internet0014merge_df, x="Year", y="Internet_Users_Pct")



In [ ]:
## 19. For how many countries was the percentage of internet users below 5% in 2014?

In [280]:
#### dataset for internet users<5% in 2014
internet_df.Internet_Users_Pct = internet_df.Internet_Users_Pct.astype(int)
internet_df['Internet_Users_Pct'<5.0 and 'Year'==2014]
#### Count number of countries in this dataset

TypeError: '<' not supported between instances of 'str' and 'float'

In [ ]:
## 20. Merge the two DataFrames to one: pd.merge(df1,df2, on='key')
## Do this in a way that keeps all rows from each of the two DataFrames. 
## Call the new DataFrame gdp_and_internet_use. 
## Look at the first and last few rows to confirm that it merged correctly.

In [ ]:
#### Group by the 'A' column
grouped = df.groupby('A')

In [ ]:
df.groupby('Country')